### PR TITLE
create a drop ladder for an array if any value is moved out

### LINKED ISSUE #34708: [MIR] double drop with slice patterns
## Version

```
$ rustc -V
rustc 1.11.0-dev (42b7c32ac 2016-07-02)
```
## STR

``` Rust
#![feature(slice_patterns, rustc_attrs)]

struct NoisyDrop;
impl Drop for NoisyDrop {
    fn drop(&mut self) { println!("splat!"); }
}

#[rustc_mir]
fn main() {
    let [x] = [NoisyDrop];
}
```
## Expected Result

`NoisyDrop` gets dropped once
## Actual Result

`NoisyDrop` is dropped twice

### PR DESCRIPTION
**Pull request by mikhail-m1**
r? @arielb1 
first commit for fix https://github.com/rust-lang/rust/issues/34708 (note: this still handles the subslice case in a very broken manner)

### CODE REVIEW

**Comment by mikhail-m1** [comment]
working on subslice support

File: src/librustc_mir/util/elaborate_drops.rs [hunk 1]
```diff
@@ -632,18 +633,33 @@ impl<'l, 'b, 'tcx, D> DropCtxt<'l, 'b, 'tcx, D>
         loop_block
     }
 
-    fn open_drop_for_array(&mut self, ety: Ty<'tcx>) -> BasicBlock {
-        debug!("open_drop_for_array({:?})", ety);
+    fn open_drop_for_array(&mut self, ety: Ty<'tcx>, opt_size: Option<u32>) -> BasicBlock {
```
Comment (arielb1) [inline src/librustc_mir/util/elaborate_drops.rs]: I think `size` should be a `u64`?

File: src/test/run-pass/destructure-array-2.rs [hunk 2]
```diff
@@ -0,0 +1,75 @@
+// Copyright 2015 The Rust Project Developers. See the COPYRIGHT
```
Comment (arielb1) [inline src/test/run-pass/destructure-array-2.rs]: Could you add this test to `src/test/run-pass/dynamic-drop.rs` instead? That has a framework that tests leak-proofness in panics.

**Comment by arielb1** [comment]
r=me using the `dynamic_drop` framework and a u64 for array sizes.

File: src/librustc_mir/util/elaborate_drops.rs [hunk 3]
```diff
@@ -632,18 +633,36 @@ impl<'l, 'b, 'tcx, D> DropCtxt<'l, 'b, 'tcx, D>
...
 
         // if size_of::<ety>() == 0 {
         //     index_based_loop
         // } else {
         //     ptr_based_loop
         // }
 
-        let tcx = self.tcx();
+        if let Some(size) = opt_size {
+            assert!(size <= (u32::MAX as u64),
```
Comment (mikhail-m1) [inline src/librustc_mir/util/elaborate_drops.rs]: i tried compile
```
let b : [bool; 5_000_000_000] = [true; 5000000000];
let [x,..] = b;
```
but compiler fails on earlier stage with OOM

**Comment by arielb1** [comment]
I suppose we might land this first and then get subslice support working.

**Pull request merged**